### ISSUE TITLE
Incompatibility with ForwardDiff v0.5.0

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaNLSolvers/NLsolve.jl#103]
The way the autodifferentiation is done isn't compatible with FowardDiff v0.5.0. With the new tagging system, it seems that the function that is used in the configuration part may not be the same function that is later solved. I think there may be an ordering issue, i.e. build config --> function setup but now it's required that function setup --> build config. But maybe it's better just to opt out of the function tagging and just use a package-wide tag so that way it's semi-safe but doesn't require building the final function before the config.

```julia
minpack: Error During Test
  Got an exception of type ForwardDiff.ConfigMismatchError{NLsolve.##28#31{#f!#8},Array{Float64,1},0xb046287d533b082d} outside of a @test
  The provided configuration (of type ForwardDiff.JacobianConfig{ForwardDiff.Tag{Array{Float64,1},0xb046287d533b082d},Float64,2,Array{ForwardDiff.Dual{ForwardDiff.Tag{Array{Float64,1},0xb046287d533b082d},Float64,2},1}}) was constructed for a function other than the current target function. ForwardDiff cannot safely perform differentiation in this context; see the following issue for details: https://github.com/JuliaDiff/ForwardDiff.jl/issues/83. You can resolve this problem by constructing and using a configuration with the appropriate target function, e.g. `ForwardDiff.GradientConfig(NLsolve.#28, x)`
  Stacktrace:
   [1] (::NLsolve.##30#33{NLsolve.##28#31{#f!#8},Array{Float64,1},ForwardDiff.JacobianConfig{ForwardDiff.Tag{Array{Float64,1},0xb046287d533b082d},Float64,2,Array{ForwardDiff.Dual{ForwardDiff.Tag{Array{Float64,1},0xb046287d533b082d},Float64,2},1}}})(::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,2}) at C:\Users\Chris\.julia\v0.6\NLsolve\src\autodiff.jl:11
   [2] trust_region_(::NLsolve.DifferentiableMultivariateFunction, ::Array{Float64,1}, ::Float64, ::Float64, ::Int64, ::Bool, ::Bool, ::Bool, ::Float64, ::Bool) at C:\Users\Chris\.julia\v0.6\NLsolve\src\trust_region.jl:87
   [3] #nlsolve#17(::Symbol, ::Float64, ::Float64, ::Int64, ::Bool, ::Bool, ::Bool, ::Function, ::Float64, ::Bool, ::NLsolve.#nlsolve, ::NLsolve.DifferentiableMultivariateFunction, ::Array{Float64,1}) at C:\Users\Chris\.julia\v0.6\NLsolve\src\nlsolve_func_defs.jl:24
   [4] (::NLsolve.#kw##nlsolve)(::Array{Any,1}, ::NLsolve.#nlsolve, ::NLsolve.DifferentiableMultivariateFunction, ::Array{Float64,1}) at .\<missing>:0
   [5] #nlsolve#19(::Symbol, ::Float64, ::Float64, ::Int64, ::Bool, ::Bool, ::Bool, ::Function, ::Float64, ::Bool, ::Bool, ::NLsolve.#nlsolve, ::#f!#8, ::Array{Float64,1}) at C:\Users\Chris\.julia\v0.6\NLsolve\src\nlsolve_func_defs.jl:70
   [6] (::NLsolve.#kw##nlsolve)(::Array{Any,1}, ::NLsolve.#nlsolve, ::Function, ::Array{Float64,1}) at .\<missing>:0
   [7] macro expansion at .\util.jl:293 [inlined]
   [8] macro expansion at C:\Users\Chris\.julia\v0.6\NLsolve\test\minpack.jl:538 [inlined]
   [9] macro expansion at .\test.jl:856 [inlined]
   [10] anonymous at .\<missing>:?
   [11] include_from_node1(::String) at .\loading.jl:569
   [12] include(::String) at .\sysimg.jl:14
   [13] macro expansion at C:\Users\Chris\.julia\v0.6\NLsolve\test\runtests.jl:27 [inlined]
   [14] anonymous at .\<missing>:?
   [15] include_from_node1(::String) at .\loading.jl:569
   [16] include(::String) at .\sysimg.jl:14
   [17] process_options(::Base.JLOptions) at .\client.jl:305
   [18] _start() at .\client.jl:371
Test Summary: | Pass  Error  Total
minpack       |    1      1      2
ERROR: LoadError: LoadError: Some tests did not pass: 1 passed, 0 failed, 1 errored, 0 broken.
while loading C:\Users\Chris\.julia\v0.6\NLsolve\test\minpack.jl, in expression starting on line 16
while loading C:\Users\Chris\.julia\v0.6\NLsolve\test\runtests.jl, in expression starting on line 26
===============================[ ERROR: NLsolve ]===============================

failed process: Process(`'C:\Users\Chris\AppData\Local\Julia-0.6.0-rc2\bin\julia.exe' -Cx86-64 '-JC:\Users\Chris\AppData\Local\Julia-0.6.0-rc2\lib\julia\sys.dll' --compile=yes --depwarn=yes --check-bounds=yes --code-coverage=none --color=no --compilecache=yes 'C:\Users\Chris\.julia\v0.6\NLsolve\test\runtests.jl'`, ProcessExited(1)) [1]

================================================================================
```

**Comment by tkelman** [comment]
Ah, Optim already has a ForwardDiff 0.5 upper bound which is why this isn't showing up in the tests here yet.

**Comment by KristofferC** [comment]
Ref https://github.com/JuliaNLSolvers/NLsolve.jl/pull/104

**Comment by KristofferC** [comment]
Fixed by #104
